### DISCUSSION TITLE
[Question][Hw4] Return type of `cmp` function

### DISCUSSION THREAD

**Discussion by p51lee** [prosyslab-classroom/cs524-program-analysis#376] [Q&A] [answered]
Hello,

In domain.ml, I thought `bool` should be returned for the comparison result of `cmp` function in numerical domains,
but it returns type `t`:
```ocaml
val cmp : Llvm.Icmp.t -> t -> t -> t
```
What is expected to returned by `cmp Llvm.Icmp.Sgt Pos Neg`?

**Comment by KihongHeo** [comment, marked as answer]
Hi. 
- `cmp` is the abstract version of concrete comparison. Since the concrete comparison returns a boolean value (in this homework, a bool value is 0 or 1), the abstract operator returns a `t` type value.
- Then, recall the concept of best abstraction.

**Comment by p51lee** [comment]
Thank you for your explanation!
